### PR TITLE
[4.5] Stac 13325

### PR DESCRIPTION
**Pull request by aacevedoosorio**
Health Sync documentation including new consistency models concept meant for release cadence 4.5

### CODE REVIEW

File: configure/health/send-health-data.md [hunk 1]
```diff
@@ -42,128 +42,11 @@ Health is sent to the receiver API via HTTP POST and has a common JSON object fo
...
-sts health send stop urn:health:sourceId:streamId
-```
-
-{% endtab %}
-{% endtabs %}
-
-You can also send health to StackState using the [StackState CLI `health send`](../../develop/reference/cli_reference.md#sts-health-send) command.
+StackState accepts health data based on a chosen [consistency model](/configure/health/health-synchronization.md#consistency-models) which should remain consistent across the life of the stream.
+* [Repeat Snapshots JSON](/configure/health/send-health-data/repeat_snapshots.md)
+* [Repeat States JSON](/configure/health/send-health-data/repeat_states.md)
```
Comment (gggina) [inline configure/health/send-health-data.md:49]: what about transactional increments?

File: configure/health/send-health-data.md [hunk 2]
```diff
@@ -42,128 +42,11 @@ Health is sent to the receiver API via HTTP POST and has a common JSON object fo
 }
 ```
 
-## JSON property: "health"
+## Different health data configurations
```
Comment (gggina) [inline configure/health/send-health-data.md:45]: Don't like the title.
Perhaps just "Health data configurations" ?

File: configure/health/send-health-data/repeat_snapshots.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
+### JSON property: "health"
```
Comment (gggina) [inline configure/health/send-health-data/repeat_snapshots.md:1]: Requires an "## Overview" section explaining what is on the page,
Also heading should be H2

File: configure/health/send-health-data/repeat_snapshots.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+  * **sub_stream_id** - Optional. Identifier for a sub set of the stream health data. When the stream data is distributed and reported by several agents, this allows snapshot lifecycles per `sub_stream_id`
+* **check_states** - A list of check states. Each check state can have the following fields:
+  * **checkStateId** - Identifier for the check state in the external system
+  * **message** - Optional. Message to display in StackState UI. Data will be interpreted as markdown allowing to have links to the external system check that generated the external check state.
+  * **health** - One of the following StackState Health state values: `Clear`, `Deviating`, `Critical`.
+  * **topologyElementIdentifier** - Used to bind the check state to a StackState topology element.
+  * **name** - Name of the external check state.
+
+
+### Send health to StackState
```
Comment (gggina) [inline configure/health/send-health-data/repeat_snapshots.md:59]: H2

File: configure/health/send-health-data/repeat_states.md [hunk 5]
```diff
@@ -0,0 +1,123 @@
+## JSON property: "health"
```
Comment (gggina) [inline configure/health/send-health-data/repeat_states.md:1]: Remove one of these duplicated lines.
Add "## Overview" section explaining what is on the page.

File: configure/health/send-health-data/transactional_increments.md [hunk 6]
```diff
@@ -0,0 +1,134 @@
+## JSON property: "health"
```
Comment (gggina) [inline configure/health/send-health-data/transactional_increments.md:1]: Add "## Overview" section explaining what is on the page

File: configure/health/debug-health-sync.md [hunk 7]
```diff
@@ -88,18 +88,49 @@ error message
...
+{% hint style="info" %}
+A sub stream status will show the metadata related to the consistency model:
+* **Repeat Snapshots** - Show repeat interval and expiry
+* **Repeat States** - Show repeat interval and expiry
+* **Transactional Increments** - Show checkpoint offset and checkpoint batch index
+{% endhint %}
+
+
+We can as well get the substream status with matched/unmatched check states detail so we can figure out if some of our health states are not finding any topology element to attach to.
+In the example below,  `checkStateId2` is listed under `Check states with identifier which has no matching topology element`. This means that it was not possible to match the check state to a topology element with the identifier `server-2`.
```
[the remark this answers is not in the document]
Comment (craffit) [reply configure/health/debug-health-sync.md:128]: yes, i included that in the sentence

File: configure/health/debug-health-sync.md [hunk 8]
```diff
@@ -127,6 +158,16 @@ sts health delete urn:health:sourceId:streamId
 
 ```
 
+### Clear health stream errors
```
[the remark this answers is not in the document]
Comment (craffit) [reply configure/health/debug-health-sync.md:161]: i added that, see if you like it

File: configure/health/health-synchronization.md [hunk 9]
```diff
@@ -25,18 +25,46 @@ The health synchronization framework works as follows:
...
+{% endtab %}
+
+{% tab title="Transactional increments" %}
+### Overview
+Is meant to be used on streaming systems where changes are signaled through a pipeline and only those individual health state changes can be delivered to StackState. As there is no repetition of data the whole consistency is handled by making sure that the entire pipeline can mantain at least once delivery guarantees. In order to achieve this StackState requires that a checkpoint and a previous checkpoints are communicated next to the check_states so it can track that no data is missing at all, metadata as `repeat_interval` or `expire_interval` are irrelevant in this case as there is no predefined periodicity on the data.
```
[the remark this answers is not in the document]
Comment (craffit) [reply configure/health/health-synchronization.md:49]: we can chat to make this more clear

File: configure/health/send-health-data.md (hunk 1 above)
Comment (craffit) [reply configure/health/send-health-data.md:49]: good catch

File: configure/health/send-health-data.md (hunk 2 above)
Comment (craffit) [reply configure/health/send-health-data.md:45]: lets look at this together

File: SUMMARY.md [hunk 10]
```diff
@@ -171,6 +171,9 @@
 * [Health](configure/health/README.md)
   * [Health synchronization](configure/health/health-synchronization.md)
   * [Send health data over HTTP](configure/health/send-health-data.md)
+    * [Send health data for Repeat Snapshots model](configure/health/send-health-data/repeat_snapshots.md)
+    * [Send health data for Repeat States](configure/health/send-health-data/repeat_states.md)
+    * [Send health data for Transactional Increments](configure/health/send-health-data/transactional_increments.md)
```
[the remark this answers is not in the document]
Comment (craffit) [reply SUMMARY.md:176]: done

File: configure/health/debug-health-sync.md [hunk 11]
```diff
@@ -88,18 +88,49 @@ error message
...
 
+{% hint style="info" %}
+A sub stream status will show the metadata related to the consistency model:
+* **Repeat Snapshots** - Show repeat interval and expiry
+* **Repeat States** - Show repeat interval and expiry
+* **Transactional Increments** - Show checkpoint offset and checkpoint batch index
+{% endhint %}
+
+
+We can as well get the substream status with matched/unmatched check states detail so we can figure out if some of our health states are not finding any topology element to attach to.
```
[the remark this answers is not in the document]
Comment (craffit) [reply configure/health/debug-health-sync.md:127]: done

File: configure/health/debug-health-sync.md [hunk 12]
```diff
@@ -137,7 +178,8 @@ For example a `SubStreamStopWithoutStart` will be closed once the health synchro
 
 | Error| Description |
 |:---|:---|
-| **StreamMissingSubStream** | Raised when the health synchronization receives messages with a previous start snapshot in place. |
+| **StreamMissingSubStream** | Raised when the health synchronization receives messages without a previous stream setup message as `start_snapshot` or `expiry`. |
+| **StreamConsistencyModelMismatch** | Raised when a stream has been created with certain consistency model [`REPEAT_SNAPSHOTS`, `REPEAT_STATES`, `TRANSACTIONAL_INCREMENTS`] and a message that belongs to a different model is observed.
```
[the remark this answers is not in the document]
Comment (craffit) [reply configure/health/debug-health-sync.md:182]: done

File: configure/health/debug-health-sync.md [hunk 13]
```diff
@@ -146,7 +188,10 @@ For example a `SubStreamStopWithoutStart` will be closed once the health synchro
 | **SubStreamExpired** | Raised when the health synchronization stops receiving data on a particular sub stream for longer than the configured `expiry_interval_s`. In this case, the sub stream will be deleted. |
 | **SubStreamLateData** | Raised when the health synchronization does not receive a complete snapshot timely based on the established `repeat_interval_s`. |
 | **SubStreamTransformerError** | Raised when the health synchronization is unable to interpret the payload sent to the receiver. For example, "Missing required field 'name'" with payload `{"checkStateId":"checkStateId3","health":"deviating","message":"Unable to provision the device. ","topologyElementIdentifier":"server-3"}` and transformation `Default Transformation`. |
-
+| **SubStreamMissingCheckpoint** | Raised when a Transactional increments sub stream has previously already observed a checkpoint but the received message is missing the `previous_checkpoint` |
```
[the remark this answers is not in the document]
Comment (craffit) [reply configure/health/debug-health-sync.md:191]: done

File: configure/health/debug-health-sync.md [hunk 14]
```diff
@@ -146,7 +188,10 @@ For example a `SubStreamStopWithoutStart` will be closed once the health synchro
...
+| **SubStreamInvalidCheckpoint** | Raised when a Transactional increments sub stream has previously already observed a checkpoint but the received message has a `previous_checkpoint` that is not equivalent to the last observed one. |
```
[the remark this answers is not in the document]
Comment (craffit) [reply configure/health/debug-health-sync.md:192]: done

File: configure/health/debug-health-sync.md [hunk 15]
```diff
@@ -146,7 +188,10 @@ For example a `SubStreamStopWithoutStart` will be closed once the health synchro
...
+| **SubStreamOutdatedCheckpoint** | Raised when a Transactional increments sub stream has previously already observed a checkpoint but the received message has a `checkpoint` that precedes the last observed one, meaning that it's data that StackState already received. |
```
[the remark this answers is not in the document]
Comment (craffit) [reply configure/health/debug-health-sync.md:193]: done

File: configure/health/health-synchronization.md [hunk 16]
```diff
@@ -25,18 +25,46 @@ The health synchronization framework works as follows:
 
 ![Health synchronization pipeline](/.gitbook/assets/health-sync-pipeline.svg)
 
+### Consistency models
+StackState health synchronization relies on different consistency models to guarantee that the data sent from an external monitoring system matches with what StackState ingests and shows. The two supported consistency models are: `Repeat Snapshots`, `Repeat States` and `Transactional Increments`.
```
[the remark this answers is not in the document]
Comment (craffit) [reply configure/health/health-synchronization.md:29]: done

File: configure/health/health-synchronization.md [hunk 17]
```diff
@@ -25,18 +25,46 @@ The health synchronization framework works as follows:
...
+{% tabs %}
+{% tab title="Repeat snapshots" %}
+### Overview
+Relies on periodically receiving the full snapshot of how the external checks look like in the external monitoring system. StackState keeps track of the external checks on each received snapshot and decides if some external check states need to be created, updated or deleted (as if they are not present on a snapshot anymore). The [Repeat Snapshots health payload](/configure/health/send-health-data/repeat_snapshots.md) accepts specific properties to specify when a snapshot starts or stops.
```
[the remark this answers is not in the document]
Comment (craffit) [reply configure/health/health-synchronization.md:33]: done

File: configure/health/health-synchronization.md [hunk 18]
```diff
@@ -25,18 +25,46 @@ The health synchronization framework works as follows:
...
+{% endtab %}
+
+{% tab title="Repeat states" %}
+### Overview
+Relies on periodically receiving external checks from the external monitoring system. StackState keeps track of the external checks and decides if some external checks need to be created or updated. In order to delete external checks it uses a configurable expiry mechanism to delete external checks that are not observed anymore. The [Repeat States health payload](/configure/health/send-health-data/repeat_states.md) accepts specific properties to specify what the expiry configuration should be.
+
+### When to use
+This model can be used when the external monitoring system is not capable of collecting all the external check in a determined time window and the best effort is just to send the external checks as they are obtained. This model offers less control into the data as it relies on an expiry configuration to delete external checks and it might happen that elements are deleted due to barely missing the expiry timeout whcih would reflect as external checks disappearing and reappearing from StackState.
```
[the remark this answers is not in the document]
Comment (craffit) [reply configure/health/health-synchronization.md:44]: done

File: configure/health/health-synchronization.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (craffit) [reply configure/health/health-synchronization.md:49]: done

[the remark this answers is not in the document]
Comment (craffit) [reply configure/health/health-synchronization.md:49]: done

[the remark this answers is not in the document]
Comment (craffit) [reply configure/health/health-synchronization.md:49]: done

[the remark this answers is not in the document]
Comment (craffit) [reply configure/health/health-synchronization.md:49]: done

[the remark this answers is not in the document]
Comment (craffit) [reply configure/health/health-synchronization.md:49]: done

[the remark this answers is not in the document]
Comment (craffit) [reply configure/health/health-synchronization.md:49]: done

File: configure/health/health-synchronization.md [hunk 19]
```diff
@@ -25,18 +25,46 @@ The health synchronization framework works as follows:
...
+{% endtab %}
+
+{% tab title="Transactional increments" %}
+### Overview
+Is meant to be used on streaming systems where changes are signaled through a pipeline and only those individual health state changes can be delivered to StackState. As there is no repetition of data the whole consistency is handled by making sure that the entire pipeline can mantain at least once delivery guarantees. In order to achieve this StackState requires that a checkpoint and a previous checkpoints are communicated next to the check_states so it can track that no data is missing at all, metadata as `repeat_interval` or `expire_interval` are irrelevant in this case as there is no predefined periodicity on the data.
+
+### When to use
+This model can be used when the external monitoring system does not have access to the total external checks state but only works on an event based approach. Requires strict control across the whole pipeline in order to guarantee no data loss.
```
[the remark this answers is not in the document]
Comment (craffit) [reply configure/health/health-synchronization.md:52]: done

File: configure/health/send-health-data/transactional_increments.md (hunk 6 above)
Comment (craffit) [reply configure/health/send-health-data/transactional_increments.md:5]: done

File: configure/health/send-health-data/repeat_states.md (hunk 5 above)
Comment (craffit) [reply configure/health/send-health-data/repeat_states.md]: done

File: configure/health/send-health-data/repeat_snapshots.md (hunk 4 above)
Comment (craffit) [reply configure/health/send-health-data/repeat_snapshots.md]: done

File: configure/health/send-health-data/repeat_snapshots.md (hunk 3 above)
Comment (craffit) [reply configure/health/send-health-data/repeat_snapshots.md]: done

**Review by craffit — APPROVED** [review]

**Pull request merged**